### PR TITLE
Propagate error messages from connectors to fybrik application

### PR DESCRIPTION
**Pull request by shlomitk1**
Fixes https://github.com/fybrik/fybrik/issues/1984
In case of error, connectors send a json object in the response body. This object is passed as a string to the error status of the relevant asset.

### CODE REVIEW

File: pkg/connectors/policymanager/clients/policymanager_openapi.go [hunk 1]
```diff
@@ -49,9 +50,22 @@ func NewOpenAPIPolicyManager(name, connectionURL string) (PolicyManager, error)
 
 func (m *openAPIPolicyManager) GetPoliciesDecisions(in *policymanager.GetPolicyDecisionsRequest,
 	creds string) (*policymanager.GetPolicyDecisionsResponse, error) {
-	resp, _, err := m.client.DefaultApi.GetPoliciesDecisions(context.Background()).XRequestCred(creds).GetPolicyDecisionsRequest(*in).Execute()
+	printErr := fmt.Sprintf("get policies decisions from %s failed", m.name)
```
Comment (roytman) [inline pkg/connectors/policymanager/clients/policymanager_openapi.go:53]: we use the message in the error cases, might be:

```golang
printErr := func() string { return fmt.Sprintf("get policies decisions from %s failed", m.name) }

if httpResponse == nil {
		if err != nil {
			return nil, errors.Wrap(err, printErr())
		}
		return nil, errors.New(printErr())
	}
```
Syntactic sugar :-)

File: pkg/connectors/policymanager/clients/policymanager_openapi.go [hunk 2]
```diff
@@ -49,9 +50,22 @@ func NewOpenAPIPolicyManager(name, connectionURL string) (PolicyManager, error)
...
+	if httpResponse == nil {
+		if err != nil {
+			return nil, errors.Wrap(err, printErr)
+		}
+		return nil, errors.New(printErr)
+	}
+	defer httpResponse.Body.Close()
 	if err != nil {
-		return nil, errors.Wrap(err, fmt.Sprintf("get policies decisions from %s failed", m.name))
+		if bodyBytes, errRead := io.ReadAll(httpResponse.Body); errRead == nil && len(bodyBytes) > 0 {
```
Comment (roytman) [inline pkg/connectors/policymanager/clients/policymanager_openapi.go:65]: it is exactly the same as `datacatalog_openapi.go.getDetailedError`. We don't have utilities in the connectors package, but maybe create a symmetric method here too.

File: connectors/katalog/pkg/connector/connector.go [hunk 3]
```diff
@@ -45,28 +45,25 @@ func (r *Handler) getAssetInfo(c *gin.Context) {
 	// Parse request
 	var request datacatalog.GetAssetRequest
 	if err := c.ShouldBindJSON(&request); err != nil {
-		r.Log.Info().Msg(err.Error())
-		c.JSON(http.StatusBadRequest, gin.H{"error": "Error during ShouldBindJSON in getAssetInfo "})
+		r.reportError(c, http.StatusBadRequest, err.Error())
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply connectors/katalog/pkg/connector/connector.go:48]: ok

File: manager/controllers/app/fybrikapplication_controller.go [hunk 4]
```diff
@@ -629,8 +629,7 @@ func (r *FybrikApplicationReconciler) checkGovernanceActions(configEvaluatorInpu
 			r.PolicyManager, appContext, &reqAction)
 	}
 	if err != nil {
-		// extend the error with a message from the policy manager that may help to understand the reason
-		return "", errors.WithMessage(err, msg)
+		return "", err
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/fybrikapplication_controller.go:632]: changed the description

File: manager/controllers/app/fybrikapplication_controller_test.go [hunk 5]
```diff
@@ -81,6 +81,29 @@ var _ = Describe("FybrikApplication Controller", func() {
 				Expect(k8sClient.Get(context.Background(), moduleKey, fetchedModule)).To(HaveOccurred(), "Should deny access")
 			}
 		})
+		It("Test GetAssetInfo error propagation", func() {
+			connector := os.Getenv("USE_MOCKUP_CONNECTOR")
+			if len(connector) > 0 && connector != "true" {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/fybrikapplication_controller_test.go:86]: The behavior is to check the explicit value.

File: connectors/katalog/pkg/connector/connector.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (shlomitk1) [reply connectors/katalog/pkg/connector/connector.go:48]: changed the log level, added a skip frame

File: pkg/connectors/datacatalog/clients/datacatalog_openapi.go [hunk 6]
```diff
@@ -55,26 +56,37 @@ func NewOpenAPIDataCatalog(name, connectionURL string) DataCatalog {
 	}
 }
 
+// getDetailedError extends the original error with the the response body JSON if available,
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/connectors/datacatalog/clients/datacatalog_openapi.go]: Removed

File: pkg/connectors/datacatalog/clients/datacatalog_openapi.go [hunk 7]
```diff
@@ -6,6 +6,7 @@ package clients
 import (
 	"context"
 	"fmt"
+	"io"
 	"net/http"
 
 	"emperror.dev/errors"
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/connectors/datacatalog/clients/datacatalog_openapi.go:12]: removed

File: pkg/connectors/policymanager/clients/policymanager_openapi.go [hunk 8]
```diff
@@ -6,6 +6,7 @@ package clients
 import (
 	"context"
 	"fmt"
+	"io"
 
 	"emperror.dev/errors"
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/connectors/policymanager/clients/policymanager_openapi.go:13]: removed

File: pkg/connectors/policymanager/clients/policymanager_openapi.go (hunk 1 above)
Comment (shlomitk1) [reply pkg/connectors/policymanager/clients/policymanager_openapi.go]: nice! changed to func

File: pkg/connectors/policymanager/clients/policymanager_openapi.go (hunk 2 above)
Comment (shlomitk1) [reply pkg/connectors/policymanager/clients/policymanager_openapi.go]: created a method

File: pkg/connectors/datacatalog/clients/datacatalog_openapi.go [hunk 9]
```diff
@@ -55,26 +56,37 @@ func NewOpenAPIDataCatalog(name, connectionURL string) DataCatalog {
 	}
 }
 
+// getDetailedError extends the original error with the the response body JSON if available,
+// or some default message - otherwise
```
[the remark this answers is not in the document]
Comment (roytman) [reply pkg/connectors/datacatalog/clients/datacatalog_openapi.go]: extend -> extends

File: pkg/connectors/policymanager/clients/policymanager_openapi.go [hunk 10]
```diff
@@ -47,11 +49,30 @@ func NewOpenAPIPolicyManager(name, connectionURL string) (PolicyManager, error)
 	}, nil
 }
 
+// getDetailedError generates an error from the response body JSON if available,
+// otherwise it extend the base error (e.g., 400 Bad Request) with the given message string
```
Comment (roytman) [inline pkg/connectors/policymanager/clients/policymanager_openapi.go:53]: extends

File: manager/controllers/app/fybrikapplication_controller_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (roytman) [reply manager/controllers/app/fybrikapplication_controller_test.go:86]: should we just check if it equals to "false"
(the comment is not crucial)

File: pkg/connectors/policymanager/clients/policymanager_openapi.go (hunk 10 above)
Comment (shlomitk1) [reply pkg/connectors/policymanager/clients/policymanager_openapi.go]: fixed

**Review by roytman — APPROVED** [review]

**Review by roytman — APPROVED** [review]

**Pull request merged**